_adr/83/index.adoc

```diff
@@ -0,0 +1,132 @@
+title: Permission API
+status: "Draft"
+authors:
+  - "Massimiliano Ziccardi"
+tags:
+  - "kafka"
+#applies_padrs: # What PADRs does this ADR apply?
+#applies_patterns: # What APs does this ADR apply?
+---
+## Context and Problem Statement
+Deciding whether a certain operation is allowed can be a complex matter and may require interactions with different components.
+
+Consumers of the Kas Fleet Manager APIs have no way of verifying in advance whether a particular request will be successful.
+The only way to do that, is to perform the same checks that Kas Fleet Manager will do while executing the action.
+
+This brings complexities and duplications. For example, just to check whether the creation of a new kafka is possible or not, at least the following checks must be performed:
+
+. Verify that the required terms and conditions have been accepted (this requires interaction with another service)
+. Make sure you have enough quota allocated. This requires:
+.. Having knowledge of the AMS endpoints and how to deal with them
+.. Having knowledge of the AMS terms used for designing the instance types (RHOSAKTrial for eval/developer and RHOSAK for standard)
+.. Enquire the quota assigned using the AMS endpoints, loop through the resources and check that enough quota is still available for product RHOSAK and resource RHOSAK
+.. If you have no quota assigned for STANDARD instances (RHOSAK), fallback to EVAL
+.. If you have quota for STANDARD (RHOSAK), but it is 0, fallback to EVAL (RHOSAKTrial)
+.. If you have quota for STANDARD (RHOSAK) and it is greater than 0 but all the granted quota is consumed, the creation should fail
+. Verify that at least one OCD Cluster has capacity for the instance you are going to create
+
+Another point of attention is that currently, any change in the flow requires an update to all the actors/consumers who are carrying out those checks.
+
+## Goals
+* API consumers can query KAS Fleet Manager Permission API to check if they can perform a certain action and receive a response with the outcome (allowed/denied) containing all the granted resources or, eventually, the missing requirements.
+
+## Non-goals
+
+## Current architecture
+* Kas Fleet Manager currently only allows the ‘try and see if it works’ approach: if what you are trying to do fails, Kas Fleet Manager gives you some details on went wrong, but you have no way of checking in advance if what you want to do is possible
+* API Consumers (like UI) are doing the required checkss into their codebase and every time a change occurs, both Kas Fleet Manager and UI have to be updated to reflect the change
+
+## Proposed architecture
+* We need to create a new API.
+The API will be authenticated and will allow the caller to check in advance whether he can perform a certain operation:
++
+```
+/can/{verb}/{resource}
+```
++
+Output of the API will be a JSON with the following structure:
++
+```
+{
+  allowed: boolean,
+  metadata: {
+    {key string}:{value}
```

Comment:
Wonder if this could be more structured. Something like:

options: [{
      name: {string},
      values: [{
          value: {any}
          disabled: {string - reason this value is disabled.} 
      }...]
}]
quotas: [{
      name: {string},
      granted: {int},
      used: {int}
}]

Reply:
A generic return value for all the checks necessarily constraints this API to only those checks that are common to all resources. That works for the things you're concerned about _right now_ (like terms, entitlements and capacity), but are we sure we're not going to be interested in other checks in the future? For example, capacity depends on the number of streaming units requested (which is a Kafka specific concept, at least right now). So an initial "yes" from the end point might become a "no" once the user has expressed interest in 5SU rather than 1SU. Maybe we don't plan right now to make a 2nd call when they change the number of SU, but are we confident that we won't change our minds about that? By using GET we''d be adopting an API that wasn't capable later refinement to support such finer-grained logic.

Also the generic return type argument is somewhat undermined by also returning arbitrary extra metadata, which clients need to know how to handle in order to make use of.

As a straw-man alternative pattern, I could imagine re-using the resource creation endpoint, only with a different content type for the request which changed the meaning of the request (and the expected request body) to have a dry-run semantic. The request type schema would be changed to be all optional properties, thus allowing the frontend to make requests with partial information. Doing this would tend to keep the logic for the dry-run and the I-really-mean-it cases in the same place, which would be easier to keep in sync, than having a separate endpoint handled by a separate code path. It would also allow us to gracefully change the logic over time to support the finer-grained logic mentioned above.

Reply:
@tombentley That's not what the api is though for. It is not meant to ask 'Can I create a Kafka with 5SU' and return an answer `yes/no`.
It is meant to answer to `Can I create a Kafka` and answer for example: "Yes, you can create one up to 3SU'.
So I don't think the 'create' endpoint in dryrun mode would fit.
If you think the API this way, the current generic structure will be able to cover this case too.

Reply:
@ziccardi yes, I understand that the straw man is not doing exactly the same thing as is described in this ADR. It is a superset. You can ask "Can I create a Kafka" and get yes/no, and then later during the user interaction you could also ask "Can I create a Kafka in AWS US east" and get yes/no, and later still "Can I create a Kafka of 5 SU in AWS US east" and get yes/no. So it _could be_ more useful, depending on whether the UI would (eventually) make use of this functionality. It is also more strongly typed, since the requests have a schema that's directly related to the schema for instance creation, rather than having to use an arbitrary, service-specific `object` to pass extra information. It is also something that could be implemented gradually, and go as fine-grained as we need.

> So I don't think the 'create' endpoint in dryrun mode would fit.

Since it contains the "Can I create a Kafka" functionality, I don't understand why it doesn't fit. Can you explain?

> the current generic structure will be able to cover this case too.

But it does so at the expense of having a schema for describing the request `metadata` properly.

Reply:
> So it could be more useful, depending on whether the UI would (eventually) make use of this functionality.

@pmuir @jgiardino would/could this actually be of use, do you think?

Reply:
> Since it contains the "Can I create a Kafka" functionality, I don't understand why it doesn't fit. Can you explain?

Because when you invoke the `create` endpoint, you have to describe exactly what you want to create and the outcome will be about creating exactly the instance you asked (for example, a `standard` instance on AWS US East).

With the `can-I` you ask `can I create a kafka`? and it will return an answer like: "Yes, you can create `standard,eval` on AWS US East or AWS EU, with up to 3SU", so you won't need to make the subsequent calls because it already gives you all the infos about what you can do.

Reply:
Thanks for the clarification @ziccardi, that makes a lot more sense now (and actually one of your examples showed this).

If we had a dedicated endpoint for each resource and verb (rather than a single parameterized endpoint), it would allow us to have a fully typed schema for the request (and possibly response), rather than relying on the untyped-`metadata` property. That would make the API a lot more self-describing and discoverable. Wdyt?

Reply:
Updates were made since this thread started, so I think this is the relevant update to this conversation:
```
{
  allowed: boolean,
  reasons: [{
    code: string,
    description: string,
    metadata: {             // additional details
      {key string}: {value string}
    }
  }]
}
```
and then is this an example of what the `metadata` could include?
```
{
  allowed: true,
  plans: ['standard.x1'],
  quota: {
    granted: 6,
    used: 5
  }
}
```

In the context of the UI, knowing what's `granted` vs `used` is definitely helpful. 
Knowing what regions are unavailable is also helpful.
It would be great to know if a region has limited capacity, so if a region can support another instance that is size 1, but not size 2, I'd like to communicate that as shown in [this example](https://docs.google.com/presentation/d/1U4j27K3L-9Ry0GPe-fcKoU48dCPlEJsfoXJUpldAY3U/edit?pli=1#slide=id.g118b16caf36_0_83):
<img width="799" alt="image" src="https://user-images.githubusercontent.com/21063328/164558627-74269fca-fc76-485f-9cea-e5df8ec44345.png">

Reply:
@jgiardino 
```
{
  allowed: boolean,
  reasons: [{
    code: string,
    description: string,
    metadata: {             // additional details
      {key string}: {value string}
    }
  }]
}
```
is what ALL new endpoints (verb/resources pair) will return. Then each of them will have its own schema adding other details.
For example, the `create/kafka` will add the `plans` and `quota`.
The `metadata` is only in the errors. See [here](https://github.com/bf2fc6cc711aee1a0c2a/architecture/pull/52/files#diff-01aa0bf716b701e03b5bacdf09f7b0deab40d658c388fa3e272db1b6255f5cd5R115) for an example.

> It would be great to know if a region has limited capacity, so if a region can support another instance that is size 1, but not size 2, I'd like to communicate that as shown in [this example](https://docs.google.com/presentation/d/1U4j27K3L-9Ry0GPe-fcKoU48dCPlEJsfoXJUpldAY3U/edit?pli=1#slide=id.g118b16caf36_0_83):

Totally agree.

Reply:
Thanks for the clarification.